modules/nlu/src/backend/application/bot/predictor.ts

```diff
@@ -0,0 +1,96 @@
+import Bluebird from 'bluebird'
+import * as sdk from 'botpress/sdk'
+import _ from 'lodash'
+import { ModelEntryService } from '../model-entry'
+import { NLUClient } from '../nlu-client'
+import { mapPredictOutput } from '../nlu-client/api-mapper'
+import { BotDefinition } from '../typings'
+
+type EventUnderstanding = Omit<sdk.IO.EventUnderstanding, 'includedContexts' | 'ms'>
+
+const isDefined = <T>(x: T | undefined): x is T => _.negate(_.isUndefined)(x)
+
+export class Predictor {
+  private _botId: string
+  private _languages: string[]
+  private _defaultLanguage: string
+
+  constructor(
+    bot: BotDefinition,
+    private _nluClient: NLUClient,
+    private _modelEntryService: ModelEntryService,
+    private _logger: sdk.Logger
+  ) {
+    this._botId = bot.botId
+    this._defaultLanguage = bot.defaultLanguage
+    this._languages = bot.languages
+  }
+
+  public predict = async (textInput: string, anticipatedLanguage?: string): Promise<EventUnderstanding> => {
+    const allModels = await Bluebird.map(this._languages, l =>
+      this._modelEntryService.get({ botId: this._botId, language: l })
+    )
+
+    const models = allModels.filter(isDefined)
+
+    let detectedLanguage: string | undefined
+    try {
+      detectedLanguage = await this._nluClient.detectLanguage(
```

Comment:
Can you remind me why we need to provide a list of model ids to language identification ?

Reply:
domain specific tokens (I think we fallback on custom lang id when fasttext results arent good enough)